include/nlohmann/json.hpp

```diff
@@ -2433,15 +2442,24 @@ class basic_json // NOLINT(cppcoreguidelines-special-member-functions,hicpp-spec
 
     /// @brief remove element from a JSON object given a key
     /// @sa https://json.nlohmann.me/api/basic_json/erase/
-    size_type erase(const typename object_t::key_type& key)
+    template < class KeyT, typename detail::enable_if_t <
+                   detail::is_usable_as_key_type<basic_json_t, KeyT>::value, int> = 0>
+    size_type erase(const KeyT& key)
     {
         // this erase only works for objects
-        if (JSON_HEDLEY_LIKELY(is_object()))
+        if (JSON_HEDLEY_UNLIKELY(!is_object()))
+        {
+            JSON_THROW(type_error::create(307, detail::concat("cannot use erase() with ", type_name()), this));
+        }
+
+        const auto it = m_value.object->find(key);
+        if (it != m_value.object->end())
         {
-            return m_value.object->erase(key);
+            m_value.object->erase(it);
+            return 1;
```

Comment:
This may be really nitpicky. The documentation of the return value states

> Number of elements removed. If ObjectType is the default std::map type, the return value will always be 0 (key was not found) or 1 (key was found).

I am not sure whether there is a relevant specialization of the ObjectType (`std::multimap`?) where any other values make sense. If not, then the wording of the documentation should be adjusted to not make it sound as if the type has any influence there.

Reply:
Now that this is guaranteed to always be 0 or 1, it makes sense to update the documentation.

Reply:
`std::multimap` isn't usable because it lacks `operator[]`. If it had `operator[]` (and possibly other missing functions) would it work? I.e., are there checks for duplicate keys in the serialization/deserialization code paths?

I'm leaning towards the function returning what the `erase()` call returns.

Reply:
Yes, this is what the previous code did:

```cpp
return m_value.object->erase(key);
```

Reply:
I don't know if you changed that or I did but, thinking about it, I believe the previous behavior is preferable.

Reply:
Turns out there was a reason for this code. The required `std::map::erase()` overload is a recent addition to the standard and so we need to use the iterator based `erase()` which doesn't return the number of erased elements.

`erase_internal()` handles this via SFINAE.